### ISSUE TITLE
C++ App to classify an Image with Pre-Trained ResNet Model using CNTK Evaluation Interface(Microsoft CNTK Eval CPUOnly-mkl NuGet)

### ISSUE THREAD

**Issue by jhalakpatel** [Microsoft/CNTK#805]
Hi,

I am trying write a test application in C++ which uses CNTK Evaluation Interface (Microsoft CNTK Eval CPUOnly-mkl NuGet)

The  overall application flow looks like this:

**Image PreProcessing**
Read_Image --> Resize_Image(224x224) --> Reshape_Image(from 3 Channel to 1 Channel i.e. 224x224 ==> 224x224x3) --> Read_Mean_File --> Subtract Mean from Flattened Vector i.e. 224x224x3

**Load the Pre-Trained Mode i.e. ResNet_34**

**Evaluate Model**

The output node value are generated for 1000 classes. 

**Issues:**

1. Does the Image pre-processing flow looks correct? Do I have to convert the image from R,G,B to BBB, GGG. RRR format as [issue 160](https://github.com/Microsoft/CNTK/issues/160). Also as per the same issue, do i need not subtract mean as the evaluation flow will take care of it.

2. Does the values of the output node has direct correspondence with the identified class. Thus output node number == class number. and the identified class is the one with highest value?

Thanks for your help. Please feel free to suggest additional inputs if I am missing something. I am new to CNTK evaluation interface usage.

**Comment by pkranen** [comment]
Regarding issue 1: The testing block in the ResNet_34 configuration uses a center crop instead of resizing the entire image to 224x224. The image layout for the ResNet model is also "cudnn" so please follow  issue 160 as you mentioned (see also ParallelExtractCHW extension in Examples\Evaluation\CSEvalClient\Program.cs). In your case I'd say you do need to subtract the mean since for ResNet_34 it was done in the reader and not in the ndl (I assume you left the pretrained model unchanged).

Regarding issue 2: Yes, output node number == class number, i.e. corresponding numbers provided in the train_map and val_map files. The identified class is the one with the highest output, correct. Note that the output values that you get are not passed through a softmax, i.e. they do not represent probabilies that sum up to 1.

**Comment by n17s** [comment]
Closing this. Feel free to reopen if necessary.

**Comment by jhalakpatel** [comment]
Thanks for the response. I was able to get the app working. I had a C++ App making call to C# Image Transformation APIs which used Resize and ParallelExtractCHW functions. 
But I now I am trying to replace the C++ console with a UWP app.
Approach 1: 
I have the UWP which calls the C# Image Transformation APIs (I have a WinRT Wrapper in between to call the managed C# code from the UWP app).  As with the UWP, I can use .Net Framework APIs such as System.Drawing.dll which is used for image preprocessing code. Is it possible to have the .Net Portable or .Net Core version of the pre-processing functions?

Approach 2:
I have the UWP app which can direct use Native EvalDLL.dll. Is it possible to get the image processing function which can be used in the UWP app?

I would prefer the solution to second approach as it removed the C# WinRT wrapper overhead.

Thanks in advance.

**Comment by zhouwangzw** [comment]
@jhalakpatel Would OpenCV be an option for you? CNTK Image Reader is using OpenCV in C++.
